### ISSUE TITLE
WeldJoint constructor is unavailable from Matlab

### ISSUE THREAD

**Issue by tkuchida** [opensim-org/opensim-core#1406]
According to the doxygen and [WeldJoint.h L48](https://github.com/opensim-org/opensim-core/blob/master/OpenSim/Simulation/SimbodyEngine/WeldJoint.h#L48), WeldJoint uses Joint's constructors. However, the following comment appears in [WeldJoint.cpp (L57)](https://github.com/opensim-org/opensim-core/blob/master/OpenSim/Simulation/SimbodyEngine/WeldJoint.cpp#L57):
```cpp
// Weld does not include reverse option since it has no coordinates to define
```
The doxygen for WeldJoint indicates that it inherits (all) constructors from Joint (as listed under "Additional Inherited Methods / Public Member Functions inherited from OpenSim::Joint"), among which is a constructor with a trailing "reverse" option (which _cannot_ be used to create a WeldJoint).

**Possible solutions:**
1. Add a constructor to WeldJoint that includes the trailing "reverse" option, and either
(a) throws an exception with a user-friendly error message, or
(b) displays a warning message and then calls the existing constructor.
2. Add a comment to the doxygen for WeldJoint indicating that it inherits all of Joint's constructors except the one with the trailing "reverse" option (potentially messy/confusing to include this level of detail in the description).

Found by @jimmyDunne :rugby_football:

**Comment by aseth1** [comment]
I am not sure I understand the issue here. The WeldJoint does inherit its constructors from Joint including the one with the "reverse" option. The comment is inside the `createMobilizedBody<T>` helper (specialized for the a Weld mobilizer) and refers to the Weld mobilizer having no reverse option since it is meaningless to reverse a Weld (so there is no need to check if it is reversed since the result is identical). The reverse option preserves of the coordinate as  a P->B relationship and t eventhough the tree is built B->P. In the case of the Weld there is no coordinate so the MobilizedBody::Weld has no option to reverse IIRC.

Is the main issue that the WeldJoint has an unnecessary optional "reverse" argument?

**Comment by tkuchida** [comment]
> The WeldJoint does inherit its constructors from Joint including the one with the "reverse" option.

I didn't try it in C++, but the constructor with the trailing "reverse" argument is not available in Matlab, as confirmed by @jimmyDunne. [WeldJoint.cpp](https://github.com/opensim-org/opensim-core/blob/master/OpenSim/Simulation/SimbodyEngine/WeldJoint.cpp#L46) specializes the constructor _without_ the trailing "reverse" argument but there is no specialization _with_ the trailing "reverse" argument—which seemed to support the hypothesis that the constructor with the trailing "reverse" argument is simply unavailable for WeldJoint. The issue is that the doxygen claims that WeldJoint inherits all of Joint's constructors, which does not appear to be true based on the code and based on the behavior in Matlab.

**Comment by aseth1** [comment]
>  WeldJoint.cpp specializes the constructor without the trailing "reverse" argument but there is no specialization with the trailing "reverse" argument.

`WeldJoint` does not specialize Joint constructors. All constructors are inherited. It does specialize the `createMobilizedBody<T>` helper (which helps to create the underlying Weld mobilizer) used within `extendAddToSystem()` but this has no impact on constructors.  If the `reverse` flag is being dropped then that may be due to SWIG wrapping.

**Comment by tkuchida** [comment]
> It does specialize the `createMobilizedBody<T>` helper (which helps to create the underlying Weld mobilizer) used within `extendAddToSystem()` but this has no impact on constructors.

Ah, ok thanks for clarifying. I didn't realize `template <> SimTK::MobilizedBody::Weld Joint::createMobilizedBody<SimTK::MobilizedBody::Weld>` was being called by the inherited constructor (I was also confused because I read `...::Weld Joint::...` as `...::WeldJoint:...` 🙀

**Comment by aseth1** [comment]
Check if  SWIG macro is applied to WeldJoint

**Comment by chrisdembia** [comment]
None of the joints take `reverse` in MATLAB/Python; see [here](https://github.com/opensim-org/opensim-core/blob/master/Bindings/simulation.i#L274):

```
    NAME(const std::string& name,
         const PhysicalFrame& parent,
         const SimTK::Vec3& locationInParent,
         const SimTK::Vec3& orientationInParent,
         const PhysicalFrame& child,
         const SimTK::Vec3& locationInChild,
         const SimTK::Vec3& orientationInChild) {
        return new NAME(name, parent, locationInParent, orientationInParent,
                    child, locationInChild, orientationInChild, false);
    }
```

**My preference** is to add `reverse` as a default argument for now so that the interface is consistent, and then to document that it will be removed in the future.

**Comment by tkuchida** [comment]
Fixed by #1567
![](http://gifrific.com/wp-content/uploads/2012/04/Tina-Fey-giving-herself-high-five.gif)
